ppai/weather: add weather forecasting sample

davidcavazos opened the pull request:
## Description

This is an end-to-end sample to train a PyTorch model in Google Cloud. It uses Dataflow to create the datasets using Earth Engine data, then trains a PyTorch model in Vertex AI with GPUs, deploys the model to Cloud Run and gets predictions.

To run while staged:

```sh
!git clone -b ppai-weather-forecasting https://github.com/davidcavazos/python-docs-samples.git
%cd python-docs-samples/people-and-planet-ai/weather-forecasting
```

## Checklist
- [ ] I have followed [Sample Guidelines from AUTHORING_GUIDE.MD](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md)
- [ ] README is updated to include [all relevant information](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#readme-file)
- [ ] **Tests** pass:   `nox -s py-3.9` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [ ] **Lint** pass:   `nox -s lint` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [ ] Please **merge** this PR for me once it is approved.
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample

jackwotherspoon commented on people-and-planet-ai/weather-forecasting/README.md:
```diff
@@ -0,0 +1,16 @@
+## 🌦 Weather forecasting -- _timeseries regression_
+
+> [![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/GoogleCloudPlatform/python-docs-samples/blob/main/people-and-planet-ai/weather-forecasting/README.ipynb)
```
in the "Before you Begin" section of the notebook itself there is a `repo_url` that points to your fork of the repo. Should this be updated to point to this samples repo?

davidcavazos replied:
Yeah, the notebook assumes the sample is already merged. At the PR description I made a note to run this cell instead to get this branch.

```
!git clone -b ppai-weather-forecasting https://github.com/davidcavazos/python-docs-samples.git
%cd python-docs-samples/people-and-planet-ai/weather-forecasting
```

jackwotherspoon reviewed (changes requested):
LGTM, couple small nits such as do we need to update the header years to 2023 now that it is the new year?

telpirion commented on people-and-planet-ai/weather-forecasting/serving/data.py:
```diff
@@ -0,0 +1,232 @@
+def get_inputs_patch(date: datetime, point: tuple, patch_size: int) -> np.ndarray:
+    """Gets the patch of pixels for the inputs.
+
+    args:
```
NIT: Does this need to have initial caps?

davidcavazos replied:
Good catch! Typo.

telpirion commented on people-and-planet-ai/weather-forecasting/serving/data.py:
```diff
@@ -0,0 +1,232 @@
+def get_labels_patch(date: datetime, point: tuple, patch_size: int) -> np.ndarray:
+    """Gets the patch of pixels for the labels.
+
+    args:
```
Initial caps?

davidcavazos replied:
Fixed

telpirion commented on people-and-planet-ai/weather-forecasting/trainer/task.py:
```diff
@@ -0,0 +1,312 @@
+from datasets.dataset_dict import DatasetDict
+import numpy as np
+import torch
+from transformers import (
```
I didn't realize this was a HuggingFace model!

davidcavazos replied:
Yeah, both `datasets` and `transformers` are HuggingFace, even if they don't "brand" their modules.

telpirion commented on people-and-planet-ai/weather-forecasting/trainer/task.py:
```diff
@@ -0,0 +1,312 @@
+    trainer.save_model(model_path)
+
+
+if __name__ == "__main__":
```
Same as previous. If you don't need the CLI set up, I recommend removing.

davidcavazos replied:
Isn't a CLI is required for a trainer script in Vertex AI? Please let me know if there's a better or recommended way to launch a custom Vertex AI training job.

telpirion commented on people-and-planet-ai/weather-forecasting/visualize.py:
```diff
@@ -0,0 +1,137 @@
+
+def render_elevation(patch: np.ndarray) -> np.ndarray:
+    palette = [
+        "000000",
```
NIT: I suggest leaving comments to describe what colors each of these is.

davidcavazos replied:
Great suggestion, I'm adding these.

davidcavazos commented on people-and-planet-ai/weather-forecasting/noxfile_config.py:
```diff
@@ -0,0 +1,43 @@
+# Copyright 2021 Google LLC
```
Thanks, I noticed there were other headers out of date and went ahead and updated those as well.

davidcavazos commented on people-and-planet-ai/weather-forecasting/create_dataset.py:
```diff
@@ -0,0 +1,233 @@
+    integers within a predifined range. Each integer value is treated
+    as a different classification.
+
+    From analyzing the precipitation data, most values are within 0 and 30,
```
It is mm/hr, I'll add that into the description. Great point!

davidcavazos commented on people-and-planet-ai/weather-forecasting/create_dataset.py:
```diff
@@ -0,0 +1,233 @@
+    with FileSystems.create(filename) as f:
+        inputs = [x for (x, _) in batch]
+        labels = [y for (_, y) in batch]
+        np.savez_compressed(f, inputs=inputs, labels=labels)
```
It is awkwardly named `savez_compressed`, although weirdly there's also a `savez` and a `save` function, all doing very similar things, but not quite.

https://numpy.org/doc/stable/reference/generated/numpy.savez_compressed.html

davidcavazos commented on people-and-planet-ai/weather-forecasting/create_dataset.py:
```diff
@@ -0,0 +1,233 @@
+        max_num_workers=max_requests,  # distributed runners
+        direct_num_workers=max(max_requests, MAX_REQUESTS),  # direct runner
+    )
+    with beam.Pipeline(options=beam_options) as pipeline:
```
That's awesome! I also think the Beam team made a great use of the `|` operator, just like unix pipes, but it can also confusing for some people. I personally like it :)

davidcavazos commented on people-and-planet-ai/weather-forecasting/create_dataset.py:
```diff
@@ -0,0 +1,233 @@
+        )
+
+
+if __name__ == "__main__":
```
Sadly, in this case, this is the recommended way of doing Apache Beam pipelines to be able to run them in multiple runners like Dataflow.

davidcavazos commented:
I am working on adding Cloud Run predictions. The code itself remains pretty much the same, but I had to refactor the directory structures due to Python packaging and how Dataflow and Vertex AI expect files.

davidcavazos commented:
I refactored a lot of the directory structure to be able to share the local packages with the appropriate stages of the sample. I also had to split the notebook into multiple notebooks, and make each test independent because there are dependency version incompatibilities between Apache Beam and HuggingFace Datasets (https://github.com/apache/beam/issues/24458).

Overall, the code itself stayed pretty much the same, just some of the files changed place and added a couple extra steps like building the local packages.

davidcavazos commented:
Tests are passing locally, although there are currently issues with the testing infrastructure.

davidcavazos commented:
`HTTPError: 503 Server Error: Service Unavailable` -- transient error, retesting

dandhlee commented on people-and-planet-ai/weather-forecasting/serving/main.py:
```diff
@@ -0,0 +1,58 @@
+
+
+def to_bool(x: str) -> bool:
+    return True if x.lower() == "true" else False
```
```suggestion
    return x.lower() == "true"
```

dandhlee commented on people-and-planet-ai/weather-forecasting/serving/main.py:
```diff
@@ -0,0 +1,58 @@
+
+@app.route("/")
+def ping() -> dict:
+    """Check that we can communicate with the service and get arguments."""
```
Docstrings should be in descriptive style, inline comments are fine to be imperative.
```suggestion
    """Checks that we can communicate with the service and get arguments."""
```

dandhlee commented on people-and-planet-ai/weather-forecasting/serving/weather-model/weather/trainer.py:
```diff
@@ -0,0 +1,181 @@
+
+
+def augmented(dataset: Dataset) -> Dataset:
+    """Augment dataset by rotating and flipping the examples."""
```
```suggestion
    """Augments dataset by rotating and flipping the examples."""
```

dandhlee commented on people-and-planet-ai/weather-forecasting/visualize.py:
```diff
@@ -0,0 +1,140 @@
+from __future__ import annotations
+
+import numpy as np
+import plotly.graph_objects as go
```
Not required for this PR (since it's time sensitive) but consider using `graph_object` rather than the shortened qualifier for better readability in the future.

davidcavazos replied:
We still need to wait until the tests run so I just went ahead and changed it. The formatting introduced more line breaks to keep lines short, but it's not a big deal. The linter seems to be happy with the new changes.

kweinmeister commented:
FYI, I just set the "do not merge" label because that build check was stuck. Now it's green.
